File: low-level-design/3-creational-design-patterns/3-builder-pattern.py
# Bad Example
# Represents a customizable Burger Meal
class BurgerMeal:
    # Constructor trying to handle all combinations
    def __init__(self, bun, patty, sides=None, toppings=None, cheese=False):
        # Mandatory components
        self.bun = bun
        self.patty = patty

        # Optional components
        self.sides = sides
        self.toppings = toppings
        self.cheese = cheese

# Constructing the object with only required details
burger_meal = BurgerMeal("wheat", "veg", None, None, False)

# Represents a customizable Burger Meal
class BurgerMealBuidler:
    # Private constructor to force use of Builder
    def __init__(self, builder):
        # Required components
        self.bunType = builder.bunType
        self.patty = builder.patty

        # Optional components
        self.hasCheese = builder.hasCheese
        self.toppings = builder.toppings
        self.side = builder.side
        self.drink = builder.drink

    # Static nested Builder class
    class BurgerBuilder:
        # Builder constructor with required fields
        def __init__(self, bunType, patty):
            # Required
            self.bunType = bunType
            self.patty = patty

            # Optional
            self.hasCheese = False
            self.toppings = []
            self.side = None
            self.drink = None

        # Method to set cheese
        def withCheese(self, hasCheese):
            self.hasCheese = hasCheese
            return self

        # Method to set toppings
        def withToppings(self, toppings):
            self.toppings = toppings
            return self

        # Method to set side
        def withSide(self, side):
            self.side = side
            return self

        # Method to set drink
        def withDrink(self, drink):
            self.drink = drink
            return self

        # Final build method
        def build(self):
            return BurgerMeal(self)

# Creating burger with only required fields
plain_burger = BurgerMealBuidler.BurgerBuilder("wheat", "veg").build()

# Burger with cheese only
burger_with_cheese = (
    BurgerMealBuidler.BurgerBuilder("wheat", "veg")
    .withCheese(True)
    .build()
)

# Fully loaded burger
toppings = ["lettuce", "onion", "jalapeno"]
loaded_burger = (
    BurgerMealBuidler.BurgerBuilder("multigrain", "chicken")
    .withCheese(True)
    .withToppings(toppings)
    .withSide("fries")
    .withDrink("coke")
    .build()
)

File: low-level-design/3-creational-design-patterns/3-builder-pattern.java
import java.util.*;

class BurgerMeal {
    private String bun;
    private String patty;

    private String sides;
    private List<String> toppings;
    private boolean cheese;

    public BurgerMeal(String bun, String patty, String sides, List<String> toppings, boolean cheese) {
        this.bun = bun;
        this.patty = patty;
        this.sides = sides;
        this.toppings = toppings;
        this.cheese = cheese;
    }

    // Telescoping Constructor Pattern
    // public BurgerMeal(String bun, String patty, boolean cheese) {}
    // public BurgerMeal(String bun, String patty, String sides, boolean cheese) {}
}

# Builder Pattern
class BurgerMealBuilder {
    // Required Parameters
    private final String bunType;
    private final String patty;

    // Optional Parameters
    private final boolean hasCheese;
    private final List<String> toppings;
    private final String side;
    private final String drink;

    private BurgerMealBuilder(BurgerBuilder builder) {
        this.bunType = builder.bunType;
        this.patty = builder.patty;
        this.hasCheese = builder.hasCheese;
        this.toppings = builder.toppings;
        this.side = builder.side;
        this.drink = builder.drink;
    }

    public static class BurgerBuilder {
        // Required Parameters
        private final String bunType;
        private final String patty;

        // Optional Parameters
        private final boolean hasCheese;
        private final List<String> toppings;
        private final String side;
        private final String drink;
        
        public BurgerBuilder(String bunType, String patty) {
            this.bunType = bunType;
            this.patty = patty;
        }

        public BurgerBuilder withCheese(boolean hasCheese) {
            this.hasCheese = hasCheese;
            return this;
        }

        public BurgerBuilder withToppings(List<String> toppings) {
            this.toppings = toppings;
            return this;
        }

        public BurgerBuilder withSide(String side) {
            this.side = side;
            return this;
        }

        public BurgerBuilder withDrink(String drink) {
            this.drink = drink;
            return this;
        }

        public BurgerMealBuilder build() {
            return new BurgerMealBuilder(this);
        }
    }
}

class Main {
    public static void main(String[] args) {
        BurgerMeal burgerMeal = new BurgerMeal("Sesame Bun", "Beef Patty", null, null, false);

        BurgerMealBuilder plainBurger = new BurgerMealBuilder.BurgerBuilder("wheat", "veg").build();
        BurgerMealBuilder cheeseBurger = new BurgerMealBuilder.BurgerBuilder("wheat", "beef").withCheese(true).build();
        BurgerMealBuilder deluxeBurger = new BurgerMealBuilder.BurgerBuilder("wheat", "beef").withCheese(true).withToppings(Arrays.asList("lettuce", "tomato", "onion")).withSide("fries").withDrink("coke").build();
    }
}